### PR TITLE
DOCS-476: OIDC, Keycloak, Dex, and Policy Variables

### PR DESCRIPTION
**Pull request by ravindk89**
Staged: 

- http://192.241.195.202:9000/staging/DOCS-476/linux/html/operations/external-iam/configure-keycloak-identity-management.html

- http://192.241.195.202:9000/staging/DOCS-476/k8s/html/operations/external-iam/configure-keycloak-identity-management.html

- http://192.241.195.202:9000/staging/DOCS-476/container/html/operations/external-iam/configure-keycloak-identity-management.html

- http://192.241.195.202:9000/staging/DOCS-476/linux/html/administration/identity-access-management/policy-based-access-control.html#policy-variables

Partially addresses #476 

Done

- [x] Keycloak on Linux + Admin API
- [x] Keycloak-specific variables added to envvar/conf
- [x] Initial setup of Dex pages
- [x] Scoping of Kubernetes / Containerization pages
- [x] Keycloak for K8s, Containers
- [x] Policy Variable load-up
- [x] Policy Variable tutorial for OIDC 

ToDo

- [ ] Dex
- [ ] Pluggable IAM

### CODE REVIEW

File: source/operations/external-iam/configure-dex-identity-management.rst [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+   - Create policies to control access of Dex-authenticated users
+   - Log into the MinIO Tenant Console using SSO and a Dex-managed identity
+   - Generate temporary S3 access credentials using the ``AssumeRoleWithWebIdentity`` Security Token Service (STS) API
+
+.. cond:: not k8s
+
+   - Configuring Dex for use with MinIO authentication and authorization
+   - Configuring a new or existing MinIO cluster to use Dex as the OIDC provider
+   - Create policies to control access of Dex-authenticated users
+   - Log into the MinIO Tenant Console using SSO and a Dex-managed identity
```
Comment (djwfyi) [inline source/operations/external-iam/configure-dex-identity-management.rst:34]: I think this should be `Log in to` (and in other files)

File: source/operations/external-iam/configure-dex-identity-management.rst [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+   - Generate temporary S3 access credentials using the ``AssumeRoleWithWebIdentity`` Security Token Service (STS) API
+
+
+This procedure was written and tested against Dex VERSION.
```
Comment (djwfyi) [inline source/operations/external-iam/configure-dex-identity-management.rst:38]: Assuming this is still in progress?

File: source/operations/external-iam/configure-keycloak-identity-management.rst [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+   .. code-block:: shell
+      :class: copyable
+
+      curl -H "Authentication: Bearer ACCESS_TOKEN_VALUE" \
+           http://keycloak-url:port/admin/realms/REALM/users/UUID
+
+   Replace the UUID with the unique ID for the user which you want to retrieve.
+   The response should resemble the following:
+
+   .. code-block:: shell
```
Comment (djwfyi) [inline source/operations/external-iam/configure-keycloak-identity-management.rst:179]: It's `JSON`, not `shell`.

File: source/operations/external-iam/configure-keycloak-identity-management.rst [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+      - Specify the Keycloak admin URL to :mc-conf:`keycloak_admin_url <identity_openid.keycloak_admin_url>` configuration setting
+
+      - Specify the Keycloak Realm name to :mc-conf:`keycloak_realm <identity_openid.keycloak_realm>`
+
+   .. tab-item:: Environment Variables
+
+      Set the following :ref:`environment variables <minio-server-envvar-external-identity-management-openid>` in the appropriate configuration location, such as ``/etc/default/minio``.
+
+      The following example code sets the minimum required environment variables related to enabling the Keycloak Admin API for an existing Keycloak configuration.
+      Replace the suffix ``_KEYCLOAK`` with the unique identifier for the target Keycloak configuration.
```
Comment (djwfyi) [inline source/operations/external-iam/configure-keycloak-identity-management.rst:247]: I'm not clear what this sentence means.
So `MINIO_IDENTITY_OPENID_VENDOR_KEYCLOAK="keycloak"` might become
`MINIO_IDENTITY_OPENID_VENDOR_MYKEYTARGET="keycloak"`?

File: source/includes/k8s/steps-configure-keycloak-identity-management.rst [hunk 5]
```diff
@@ -0,0 +1,283 @@
...
+
+- Using the MinIO Operator Console
+- Using the MinIO Tenant Console
+- Using a terminal/shell and the :mc:`mc admin idp openid` command
+
+.. tab-set::
+
+   .. tab-item:: MinIO Operator Console
+
+      You can use the MinIO Operator Console to configure Keycloak as the External Identity Provider for the MinIO Tenant:
```
Comment (djwfyi) [inline source/includes/k8s/steps-configure-keycloak-identity-management.rst:101]: The way this ends with the `:` just looks odd when rendered.
Swap to a period maybe?

With the tabs included for the port-forward/node ports include, it looks like the whole section is left undone.

File: source/includes/k8s/steps-configure-keycloak-identity-management.rst [hunk 6]
```diff
@@ -0,0 +1,283 @@
...
+
+      .. include:: /includes/common/common-k8s-connect-operator-console.rst
```
Comment (djwfyi) [inline source/includes/k8s/steps-configure-keycloak-identity-management.rst:103]: I'm also not a fan of how you have tabs nested under the MinIO Operator Console tab. It's sort of visually confusing. Not sure what to do here, since it's an include.
My first thought was to rework this section as dropdowns instead of tabs, but not sure that impacts all the other places you use this content.

image

File: source/operations/external-iam/configure-keycloak-identity-management.rst [hunk 7]
```diff
@@ -0,0 +1,259 @@
...
+This functionality allows MinIO to more quickly remove access from previously authenticated Keycloak users.
+
+This procedure assumes an existing MinIO deployment configured with Keycloak as an external identity manager.
+
+1) Create the Necessary Client Scopes
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Navigate to the :guilabel:`Client scopes` view and create a new scope:
+
+- Set the :guilabel:`Name` to a recognizable name for the scope (``minio-admin-API-access``)
```
Comment (djwfyi) [inline source/operations/external-iam/configure-keycloak-identity-management.rst:141]: This section renders as a mess of gray boxes.
Maybe field lists will improve it? Or a table?

image

File: source/includes/container/steps-configure-keycloak-identity-management.rst [hunk 8]
```diff
@@ -0,0 +1,324 @@
...
+          --pod minio-keycloak \
+          -e KEYCLOAK_ADMIN=keycloakadmin \
+          -e KEYCLOAK_ADMIN_PASSWORD=keycloakadmin123 \
+          quay.io/keycloak/keycloak:latest start-dev
+
+You can access the Keycloak container over the ``http://localhost:8080`` in your browser.
+
+3) Configure or Create a Client for Accessing Keycloak
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
```
Comment (djwfyi) [inline source/includes/container/steps-configure-keycloak-identity-management.rst:41]: The `guilabel` and monotype doesn't work for these lists. It's really hard to parse.
It's just a sea of gray boxes.

Field lists or tables instead?

image

**Review by djwfyi — CHANGES_REQUESTED** [review]
This is an incredible amount of work. Well done!
Some of the lists really don't work the way they currently render, so maybe try list tables or the field lists instead?

Otherwise, just a bunch of nits.

I'll take a look again when you think it needs it.

File: source/includes/k8s/steps-configure-keycloak-identity-management.rst (hunk 6 above)
Comment (ravindk89) [reply source/includes/k8s/steps-configure-keycloak-identity-management.rst:50]: Redone with dropdowns.

File: source/operations/external-iam/configure-keycloak-identity-management.rst (hunk 4 above)
Comment (ravindk89) [reply source/operations/external-iam/configure-keycloak-identity-management.rst:256]: Basically yes. Not sure how to approach this to be honest. Open to suggestions!

**Review by ravindk89 — COMMENTED** [review]
@djwfyi @feorlen OK so included Daryl's last slate of changes.

Am still working on pluggable IDP + other fixups related to ongoing work, but if this core stuff looks OK (I am going to drop dex before merge) then this procedure at least is solid.

@donatello LMK if anything looks off.

File: source/operations/external-iam/configure-keycloak-identity-management.rst [hunk 9]
```diff
@@ -0,0 +1,267 @@
...
+
+   You can also use this procedure as guidance for modifying an existing MinIO Tenant to enable Keycloak Identity Management.
+
+.. cond:: linux or container or macos or windows
+
+   MinIO Deployment
+   ~~~~~~~~~~~~~~~~
+
+   This procedure assumes an existing MinIO cluster running the :minio-git:`latest stable MinIO version <minio/releases/latest>`. 
+   Defer to the :ref:`minio-installation` for more complete documentation on new MinIO deployments.
```
Comment (djwfyi) [inline source/operations/external-iam/configure-keycloak-identity-management.rst:75]: I think `Refer` instead of `Defer` here.

File: source/operations/external-iam/configure-keycloak-identity-management.rst (hunk 4 above)
Comment (djwfyi) [reply source/operations/external-iam/configure-keycloak-identity-management.rst:256]: So I think my suggestion would be something like

```
         MINIO_IDENTITY_OPENID_VENDOR_{SUFFIX}="keycloak"
         MINIO_IDENTITY_OPENID_KEYCLOAK_ADMIN_URL_{SUFFIX}="https://keycloak-url:port/admin"
```

Then say, 
> Replace `{SUFFIX}` with a unique identifier for your keycloak IDP. For example, `MINIO_IDENTITY_OPENID_VENDOR_KEYCLOAK`. 
If you do not expect to have more than one OpenID vendor, you can leave the suffix off entirely and use `MINIO_IDENTITY_OPENID_VENDOR`.

File: source/administration/identity-access-management/policy-based-access-control.rst [hunk 10]
```diff
@@ -816,3 +816,218 @@ MinIO supports the following conditions for use with defining policies for
...
+                  "s3:GetObject",
+                  "s3:PutObject"
+                  ],
+                  "Effect": "Allow",
+                  "Resource": ["arn:aws:s3:::mybucket/${jwt:PreferredUsername}/*"]
+               }
+            ]
+         }
+
+      The ``${jwt:PreferredUsername}`` substitutes in the username to that portion of the policy.
```
Comment (feorlen) [inline source/administration/identity-access-management/policy-based-access-control.rst:985]: > substitutes in the username to that portion of the policy.
It took me a few tries to read this as intended. (There's also a similar line below, in the LDAP example.)

My understanding is these are template variables, replaced by a real value when known. Perhaps something like this?

> ``${jwt:PreferredUsername}`` in the ``Resource`` field is replaced with the authenticated user's name.

I feel like there could be "when the policy is applied" at the end too, although I'm not sure that is correct terminology.

File: source/includes/common/common-k8s-connect-operator-console.rst [hunk 11]
```diff
@@ -1,81 +1,80 @@
...
-      .. cond:: openshift
+      .. code-block:: shell
+         :class: copyable
 
-         .. code-block:: shell
-            :class: copyable
+         oc minio proxy
 
-            oc minio proxy
+   The command output includes a JWT token you must use to log into the
```
Comment (feorlen) [inline source/includes/common/common-k8s-connect-operator-console.rst:28]: Perhaps replace _you must use_ with _needed_ or _required_ here?

File: source/operations/external-iam/configure-dex-identity-management.rst [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+
+   You can also use this procedure as guidance for modifying an existing MinIO Tenant to enable Dex Identity Management.
+
+.. cond:: linux or container or macos or windows
+
+   MinIO Deployment
+   ~~~~~~~~~~~~~~~~
+
+   This procedure assumes an existing MinIO cluster running the :minio-git:`latest stable MinIO version <minio/releases/latest>`. 
+   Defer to the :ref:`minio-installation` for more complete documentation on new MinIO deployments.
```
Comment (feorlen) [inline source/operations/external-iam/configure-dex-identity-management.rst:66]: _Defer to_ feels a little fancy to me, what about _See_?

File: source/operations/external-iam/configure-keycloak-identity-management.rst [hunk 13]
```diff
@@ -0,0 +1,267 @@
...
+         "access": {
+            "manageGroupMembership": true,
+            "view": true,
+            "mapRoles": true,
+            "impersonate": true,
+            "manage": true
+         }
+      }
+
+   MinIO would revoke access for an authenticated user if the returned value has ``enabled: false``, or if the Admin API returned no matching user.
```
Comment (feorlen) [inline source/operations/external-iam/configure-keycloak-identity-management.rst:216]: I'm not entirely following this paragraph. The first part makes sense: if the returned value has `enabled: false`, access is revoked even if the user would otherwise be correctly authenticated.

Does _the Admin API returned no matching user_ mean the user doesn't exist, or the user does exist but lacks some needed permission? Is there anything to revoke in that case?

File: source/operations/external-iam/configure-keycloak-identity-management.rst [hunk 14]
```diff
@@ -0,0 +1,267 @@
...
+
+         mc admin idp openid update ALIAS KEYCLOAK_IDENTIFIER \
+            vendor="keycloak" \
+            keycloak_admin_url="https://keycloak-url:port/admin"
+            keycloak_realm="REALM"
+
+      - Replace ``KEYCLOAK_IDENTIFIER`` with the name of the configured Keycloak IDP.
+        You can use :mc-cmd:`mc admin idp openid list` to view all configured IDP configurations on the MinIO deployment
+        
+      - Specify the Keycloak admin URL to :mc-conf:`keycloak_admin_url <identity_openid.keycloak_admin_url>` configuration setting
```
Comment (feorlen) [inline source/operations/external-iam/configure-keycloak-identity-management.rst:247]: _Specify... to_ reads a bit oddly to me, although I'm not sure what might be better given I'm not familiar with the context.
(Here and also next item.)

**Review by ravindk89 — COMMENTED** [review]
@djwfyi @feorlen 

OK back to both of you. I  think that should be everything, but a bit more:

- Added the `AssumeRoleWithCustomToken` STS API + a new addition to the Administration -> Identity Management - > MinIO Identity Management Plugin

Hopefully we're close to done here. @donatello I figure you might be busy, but if you can find 30 minutes this week to look over this that would be appreciated. Otherwise we will push this up and refine later.

File: source/operations/external-iam/configure-keycloak-identity-management.rst (hunk 14 above)
Comment (feorlen) [reply source/operations/external-iam/configure-keycloak-identity-management.rst:246]: To clarify my comment, it's more about the `to` bit. "Specify _thing_ *to* _URL_."  I think I'd be more likely to write it as "specify _thing_ *for* _URL_."

Maybe that better describes what I was thinking? I'm not entirely certain what `keycloak_admin_url` is, so perhaps I'm the one misunderstanding.

File: source/includes/common/common-configure-keycloak-identity-management.rst [hunk 15]
```diff
@@ -0,0 +1,370 @@
...
+   :widths: 30 70
+   :width: 100%
+
+   * - :guilabel:`Name` 
+     - Enter a unique name for the Keycloak instance 
+   
+   * - :guilabel:`Config URL`
+     - Specify the address of the Keycloak OpenID configuration document (|KEYCLOAK_URL|)
+
+       Ensure the ``REALM`` matches the Keycloak realm you want to use for authenticating users to MinIO
```
Comment (djwfyi) [inline source/includes/common/common-configure-keycloak-identity-management.rst:238]: While the bullet shouldn't have a `.`, I would end the extra note here and in each of the below bullets with one. Especially since you use periods for the last point.

File: source/includes/common/common-k8s-connect-operator-console.rst [hunk 16]
```diff
@@ -1,81 +1,77 @@
-.. tab-set::
+.. dropdown:: Port Forwarding
+   :open:
 
-   .. tab-item:: Port Forwarding
+   .. note::
+      
+      Some Kubernetes deployments may experience issues with timeouts during port-forwarding operations with the Operator Console.
+      Select the :guilabel:`Node Ports` section to view instructions for alternative access.
```
Comment (djwfyi) [inline source/includes/common/common-k8s-connect-operator-console.rst:7]: It's `NodePorts` (no space). Probably worth doing a find/replace.

File: source/includes/common/common-k8s-connect-operator-console.rst [hunk 17]
```diff
@@ -1,81 +1,77 @@
...
-      .. cond:: openshift
+      .. code-block:: shell
+         :class: copyable
 
-         .. code-block:: shell
-            :class: copyable
+         oc minio proxy
 
-            oc minio proxy
+   The command output includes a required token required for logging into the Operator Console.
```
Comment (djwfyi) [inline source/includes/common/common-k8s-connect-operator-console.rst:27]: Drop one of the `required`. I suggest dropping the first one.

File: source/includes/k8s/steps-configure-keycloak-identity-management.rst [hunk 18]
```diff
@@ -0,0 +1,143 @@
...
+             
+             For MinIO Console deployments behind a reverse proxy, load balancer, or similar network control plane, you can instead use the :envvar:`MINIO_BROWSER_REDIRECT_URL` variable to set the redirect address for Keycloak to use.
+
+      Select :guilabel:`Save` to apply the configuration.
+
+   .. tab-item:: MinIO Tenant Console
+
+      You can use the MinIO Tenant Console to configure Keycloak as the External Identity Provider for the MinIO Tenant.
+
+      Access the Console service using the Node Port, Ingress, or Load Balancer endpoint.
```
Comment (djwfyi) [inline source/includes/k8s/steps-configure-keycloak-identity-management.rst:97]: NodePort, no space

File: source/includes/k8s/steps-configure-keycloak-identity-management.rst [hunk 19]
```diff
@@ -0,0 +1,143 @@
...
+
+      Select :guilabel:`Save` to apply the configuration.
+
+   .. tab-item:: CLI
+
+      .. include:: /includes/common/common-configure-keycloak-identity-management.rst
+         :start-after: start-configure-keycloak-minio-cli
+         :end-before: end-configure-keycloak-minio-cli
+
+You must restart the MinIO deployment for the changes to apply.
```
Comment (djwfyi) [inline source/includes/k8s/steps-configure-keycloak-identity-management.rst:119]: Maybe
> Restart the MinIO deployment to apply the changes.

File: source/operations/external-iam.rst [hunk 20]
```diff
@@ -10,14 +10,18 @@ External Identity Management
    :local:
    :depth: 2
 
-MinIO supports offloading identity management onto one of the following supported IDentity Providers (IDP):
+MinIO supports multiple external identity managers through the following supported IDentity Providers (IDP):
```
Comment (djwfyi) [inline source/operations/external-iam.rst:13]: I'd drop `supported` at the end of the line. You already said "MinIO supports" at the beginning.

File: source/operations/external-iam/configure-keycloak-identity-management.rst (hunk 14 above)
Comment (djwfyi) [reply source/operations/external-iam/configure-keycloak-identity-management.rst:246]: Maybe
> Specify the Keycload admin URL in the ...

I agree that the `to` phrasing catches me by surprise.

File: source/reference/minio-mc-admin/mc-admin-config.rst [hunk 21]
```diff
@@ -2106,7 +2106,7 @@ configuration settings.
 
       *Optional*
 
-      .. include:: /includes/common-minio-external-auth.rst
+]      .. include:: /includes/common-minio-external-auth.rst
```
Comment (djwfyi) [inline source/reference/minio-mc-admin/mc-admin-config.rst:2109]: I feel like this open bracket is out of place.

File: source/operations/external-iam/configure-keycloak-identity-management.rst [hunk 22]
```diff
@@ -0,0 +1,266 @@
...
+
+2. Use the value returned as the ``access_token`` to access the Admin API:
+
+   .. code-block:: shell
+      :class: copyable
+
+      curl -H "Authentication: Bearer ACCESS_TOKEN_VALUE" \
+           http://keycloak-url:port/admin/realms/REALM/users/UUID
+
+   Replace the UUID with the unique ID for the user which you want to retrieve.
```
Comment (djwfyi) [inline source/operations/external-iam/configure-keycloak-identity-management.rst:186]: Maybe
> Replace `UUID` with the unique ID ...

**Review by djwfyi — APPROVED** [review]
Nibbles and bits throughout, but nothing really substantive that I would need to see.
Approving it, with the expectation you have the noted fixes in place before you merge it. 😄 

Stunning amount of work in this one. 🛌

File: source/developers/security-token-service/AssumeRoleWithCustomToken.rst
[the remark this answers is not in the document]
Comment (ravindk89) [reply source/developers/security-token-service/AssumeRoleWithCustomToken.rst:1]: I checked and it is linked. I think. Maybe it wasn't in staging?

File: source/developers/security-token-service/AssumeRoleWithCustomToken.rst [hunk 23]
```diff
@@ -0,0 +1,117 @@
...
+
+       Specify ``2011-06-15``.
+
+   * - ``RoleArn``
+     - string
+     - *Required*
+
+       Specify the ARN for the Identity Manager Plugin configuration to associate with this STS request.
+
+       See :envvar:`MINIO_IDENTITY_PLUGIN_ROLE_ID` or :mc-conf:`identity_plugin role_id` for more information.
```
[the remark this answers is not in the document]
Comment (ravindk89) [reply source/developers/security-token-service/AssumeRoleWithCustomToken.rst:70]: Thought I fixed that. Will fix-fix it.

File: source/includes/common/common-configure-keycloak-identity-management.rst [hunk 24]
```diff
@@ -0,0 +1,370 @@
...
+     - Set to a unique identifier for MinIO (``minio``)
+   * - :guilabel:`Client type` 
+     - Set to ``OpenID Connect``
+   * - :guilabel:`Always display in console`
+     - Toggle to ``On``
+   * - :guilabel:`Client authentication`
+     - Toggle to ``On``
+   * - :guilabel:`Authentication flow`
+     - Toggle on ``Standard flow``
+   * - (Optional) :guilabel:`Authentication flow`
```
[the remark this answers is not in the document]
Comment (ravindk89) [reply source/includes/common/common-configure-keycloak-identity-management.rst:21]: Resolve out-of-band and fix everywhere?

File: source/includes/common/common-configure-keycloak-identity-management.rst [hunk 25]
```diff
@@ -0,0 +1,370 @@
...
+   * - :guilabel:`Token Claim Name` 
+     - Set to ``policy``
+   * - :guilabel:`Add to ID token` 
+     - Set to ``On``
+   * - :guilabel:`Claim JSON Type` 
+     - Set to ``String``
+   * - :guilabel:`Multivalued` 
+     - Set to ``On``
+
+       This allows users to inherit any ``policy`` set in their Groups
```
[the remark this answers is not in the document]
Comment (ravindk89) [reply source/includes/common/common-configure-keycloak-identity-management.rst:84]: You need both - the first one lets the policy be an array, and the second aggregates it together properly.

I think.

[the remark this answers is not in the document]
Comment (ravindk89) [reply source/includes/common/common-configure-keycloak-identity-management.rst:84]: Made a tweak

File: source/includes/common/common-configure-keycloak-identity-management.rst [hunk 26]
```diff
@@ -0,0 +1,370 @@
...
+           -d "WebIdentityToken=TOKEN"
+
+   Replace the ``TOKEN`` with the ``access_token`` value returned by Keycloak.
+
+   The API returns an XML document on success containing the following keys:
+   
+   - ``Credentials.AccessKeyId`` - the Access Key for the Keycloak User
+   - ``Credentials.SecretAccessKey`` - the Secret Key for the Keycloak User
+   - ``Credentials.SessionToken`` - the Session Token for the Keycloak User
+   - ``Credentials.Expiration`` - the Expiration Date for the generated credentials.
```
[the remark this answers is not in the document]
Comment (ravindk89) [reply source/includes/common/common-configure-keycloak-identity-management.rst:193]: Let it be none of them

File: source/reference/minio-mc-admin/mc-admin-config.rst (hunk 21 above)
Comment (ravindk89) [reply source/reference/minio-mc-admin/mc-admin-config.rst:2109]: That is a Ctrl + ] whoopsie
